ssl/statem/extensions.c

```diff
@@ -850,6 +850,19 @@ static int final_server_name(SSL *s, unsigned int context, int sent,
     }
 
     /*
+     * If we switched contexts (whether here or in the client_hello callback,
+     * move the sess_accept increment from the session_ctx to the new
+     * context, to avoid the confusing situation of having sess_accept_good
+     * exceed sess_accept (zero) for the new context.
+     */
+    if (SSL_IS_FIRST_HANDSHAKE(s) && s->ctx != s->session_ctx) {
+        CRYPTO_atomic_add(&s->ctx->stats.sess_accept, 1, &discard,
+                          s->ctx->lock);
+        CRYPTO_atomic_add(&s->session_ctx->stats.sess_accept, -1, &discard,
+                          s->session_ctx->lock);
```

Comment:
One might argue that we should also check right after the client_hello callback, since then we would not "lose" any dropped connections due to bogus extensions, etc.  I chose to only have one location where we do this, as otherwise I would need to cache the value of s->ctx around the servername_cb call, to avoid double-counting any changes made in the client_hello callback, and that seemed an overly invasive change to me.